src/extensions/step_handlers/wiot_handler/src/wiot_handler_1.cpp

```diff
@@ -0,0 +1,353 @@
+        }
+    }
+
+    return ADUCResult(ADUC_Result_Install_Success);
+}
+
+/**
+ * @brief Apply implementation
+ *
+ * @return ADUC_Result The result of the apply.
+ * ADUC_Result_Success - success
+ * ADUC_Result_Apply_RequiredImmediateReboot - reboot required
+ */
+ADUC_Result WiotHandler1::Apply(const ADUC_WorkflowData* workflowData)
+{
+    Log_Info(HANDLER_LOG_ID "Apply");
+
+    const ADUC_WorkflowHandle workflowHandle{ workflowData->WorkflowHandle };
+
+    if (workflow_get_operation_cancel_requested(workflowHandle))
+    {
+        return Cancel(workflowData);
+    }
+
+    workflow_string_ptr workFolder(workflow_get_workfolder(workflowHandle));
+    if (workFolder.get() == nullptr)
+    {
+        Log_Error("Unable to get work folder");
+
+        return ADUCResult(ADUC_Result_Failure, ADUC_ERC_SCRIPT_HANDLER_APPLY_FAILURE_UNKNOWNEXCEPTION);
+    }
+
+    Log_Info("Applying from %s", workFolder.get());
+
+    if (!StepHandler::Apply(workFolder.get()))
+    {
+        Log_Error("Apply failed");
+        return ADUCResult(ADUC_Result_Failure, ADUC_ERC_SCRIPT_HANDLER_APPLY_FAILURE_UNKNOWNEXCEPTION);
+    }
+
+    return ADUCResult(ADUC_Result_Success);
+}
+
+/**
+ * @brief Cancel implementation
+ *
+ * @return ADUC_Result The result of the cancel.
+ * ADUC_Result_Cancel_Success - success
+ * ADUC_Result_Cancel_UnableToCancel - unable to cancel
+ */
+ADUC_Result WiotHandler1::Cancel(const ADUC_WorkflowData* workflowData)
+{
+    Log_Info(HANDLER_LOG_ID "Cancel");
+
+    const ADUC_WorkflowHandle workflowHandle{ workflowData->WorkflowHandle };
+    const char* workflowId = workflow_peek_id(workflowHandle);
+    const int workflowLevel = workflow_get_level(workflowHandle);
+    const int workflowStep = workflow_get_step_index(workflowHandle);
+
+    Log_Info(
+        "Requesting cancel operation (workflow id '%s', level %d, step %d).", workflowId, workflowLevel, workflowStep);
+    if (!workflow_request_cancel(workflowHandle))
+    {
+        Log_Error(
+            "Cancellation request failed. (workflow id '%s', level %d, step %d)",
+            workflowId,
+            workflowLevel,
+            workflowStep);
+        return ADUCResult(ADUC_Result_Cancel_UnableToCancel);
+    }
+
+    return ADUCResult(ADUC_Result_Cancel_Success);
+}
+
+/**
+ * @brief Backup implementation
+ *
+ * @return ADUC_Result The result of the backup.
+ * ADUC_Result_Backup_Success - success (or no-op)
+ */
+ADUC_Result WiotHandler1::Backup(const ADUC_WorkflowData* workflowData)
+{
+    Log_Info(HANDLER_LOG_ID "Backup");
+
+    return ADUCResult(ADUC_Result_Backup_Success);
+}
+
+/**
+ * @brief Restore implementation.
+ *
+ * @return ADUC_Result The result of the restore.
+ * ADUC_Result_Restore_Success - success (or no-op)
+ */
+ADUC_Result WiotHandler1::Restore(const ADUC_WorkflowData* workflowData)
+{
+    Log_Info(HANDLER_LOG_ID "Restore");
+
+    return ADUCResult(ADUC_Result_Restore_Success);
```

Comment:
Might be an opportunity to call down into the install technology to attempt to revert/rollback to previous partition/firmware?

Reply:
yes, i'll forward this to my step_handler namespace.